gst/nnstreamer/tensor_filter/tensor_filter.c

```diff
@@ -638,242 +654,321 @@ _gst_tensor_filter_transform_validate (GstBaseTransform * trans,
-  GstTensorInfo *_info;
-
-  /* TODO: Consider malloc in_mem, in_info, etc. rather than take the stack */
-  GstMemory *in_mem[NNS_TENSOR_SIZE_LIMIT] = { 0, };
-  GstMapInfo in_info[NNS_TENSOR_SIZE_LIMIT];
-  GstMemory *out_mem[NNS_TENSOR_SIZE_LIMIT] = { 0, };
-  GstMapInfo out_info[NNS_TENSOR_SIZE_LIMIT];
-
-  GstTensorMemory in_tensors[NNS_TENSOR_SIZE_LIMIT];
-  GstTensorMemory invoke_tensors[NNS_TENSOR_SIZE_LIMIT];
-  GstTensorMemory out_tensors[NNS_TENSOR_SIZE_LIMIT];
-
-  GList *list;
-  guint i, num_tensors;
-  gint ret;
-  gboolean allocate_in_invoke, in_flexible, out_flexible;
-  gboolean need_profiling;
-  gsize expected, hsize;
-
-  GstTensorMetaInfo in_meta[NNS_TENSOR_SIZE_LIMIT];
-  GstTensorMetaInfo out_meta[NNS_TENSOR_SIZE_LIMIT];
-
-  GstMemory *mem;
-
-  /* 0. Check all properties. */
-  GstFlowReturn retval = _gst_tensor_filter_transform_validate (trans, inbuf,
-      outbuf);
-  if (retval != GST_FLOW_OK)
-    return retval;
+  guint i;
+  gsize hsize;
+  FilterTransformData *trans_data = NULL;
 
-  allocate_in_invoke = gst_tensor_filter_allocate_in_invoke (priv);
+  trans_data = g_new0 (FilterTransformData, 1);
 
-  in_flexible =
-      gst_tensor_pad_caps_is_flexible (GST_BASE_TRANSFORM_SINK_PAD (trans));
-  out_flexible =
-      gst_tensor_pad_caps_is_flexible (GST_BASE_TRANSFORM_SRC_PAD (trans));
-
-  if (priv->prop.invoke_dynamic && !out_flexible) {
-    ml_loge
-        ("Dynamic Invoke of tensor filter is activated but the output of tensor filter is static tensors. Currently, only flexible tensors is supported as output of dynamic invoke. If you don't want to dynamic invoke, remove the invoke-dynamic option of tensor filter.");
-    return GST_FLOW_ERROR;
+  if (!trans_data) {
+    ml_loge ("Failed to allocate memory for internal data of tensor filter transform input data.");
+    return NULL;
   }
 
-  /* 1. Get all input tensors from inbuf. */
-  /* Internal Logic Error or GST Bug (sinkcap changed!) */
-  num_tensors = gst_tensor_buffer_get_count (inbuf);
+  trans_data->num_tensors = gst_tensor_buffer_get_count (buf);
 
-  for (i = 0; i < num_tensors; i++) {
-    in_mem[i] = gst_tensor_buffer_get_nth_memory (inbuf, i);
-    if (!gst_memory_map (in_mem[i], &in_info[i], GST_MAP_READ)) {
+  for (i = 0; i < trans_data->num_tensors; i++) {
+    trans_data->mem[i] = gst_tensor_buffer_get_nth_memory (buf, i);
+    if (!gst_memory_map (trans_data->mem[i], &trans_data->info[i],
+            GST_MAP_READ)) {
       ml_logf_stacktrace
           ("gst_tensor_filter_transform: For the given input buffer, tensor-filter (%s : %s) cannot map input memory from the buffer for reading. The %u-th memory chunk (%u-th tensor) has failed for memory map.\n",
           prop->fwname, TF_MODELNAME (prop), i, i);
-      goto mem_map_error;
+      g_free (trans_data);
+      return NULL;
     }
 
+    trans_data->is_flexible =
+        gst_tensor_pad_caps_is_flexible (GST_BASE_TRANSFORM_SINK_PAD (trans));
```

Comment:
I think it would be a good idea to move this part above the For loop. What do you think? :smile: :

Reply:
Right. This oart doesn't nned to be repeated every time. Updated.